theano/gpuarray/roi_pool.c

```diff
@@ -0,0 +1,220 @@
+            for (ga_size w = wstart; w < wend; ++w) {
+            ga_size bottom_index = h * width + w;
+                if (bottom_data[bottom_index] > maxval) {
+                  maxval = bottom_data[bottom_index];
+                  maxidx = bottom_index;
+                }
+            }
+        }
+        top_data[index] = maxval;
+        argmax_data[index] = maxidx;
+    }
+}
+
+#kernel ROIPoolGPUBkwd_kernel : size, *, *, size, size, size, size, size, size, size, *, * :
+
+KERNEL void ROIPoolGPUBkwd_kernel(
+    const ga_size nloops, const DTYPE_i0* top_diff,
+    const DTYPE_i0* argmax_data, const ga_size num_rois, const DTYPE_i0 spatial_scale,
+    const ga_size channels, const ga_size height, const ga_size width,
+    const ga_size pooled_height, const ga_size pooled_width, DTYPE_i0* bottom_diff,
+    const DTYPE_i0* bottom_rois) {
+    for (ga_size index = 0; index < nloops; ++index) {
+        // (n, c, h, w) coords in bottom data
+        ga_size w = index % width;
+        ga_size h = (index / width) % height;
+        ga_size c = (index / width / height) % channels;
+        ga_size n = index / width / height / channels;
+
+        DTYPE_i0 gradient = 0;
+        // Accumulate gradient over all ROIs that pooled this element
+        for (ga_size roi_n = 0; roi_n < num_rois; ++roi_n) {
+            const DTYPE_i0* offset_bottom_rois = bottom_rois + roi_n * 5;
+            ga_size roi_batch_index = offset_bottom_rois[0];
+            if (n != roi_batch_index) {
+            continue;
+            }
+
+            ga_size roi_start_w = round(offset_bottom_rois[1] * spatial_scale);
+            ga_size roi_start_h = round(offset_bottom_rois[2] * spatial_scale);
+            ga_size roi_end_w = round(offset_bottom_rois[3] * spatial_scale);
+            ga_size roi_end_h = round(offset_bottom_rois[4] * spatial_scale);
+
+            // Skip if ROI doesn't include (h, w)
+            const ga_bool in_roi = (w >= roi_start_w && w <= roi_end_w &&
+                               h >= roi_start_h && h <= roi_end_h);
+            if (!in_roi) {
+            continue;
+            }
+
+            ga_size offset = (roi_n * channels + c) * pooled_height * pooled_width;
+            const ga_size* offset_top_diff = top_diff + offset;
+            const ga_size* offset_argmax_data = argmax_data + offset;
+
+            // Compute feasible set of pooled units that could have pooled
+            // this bottom unit
+
+            // Force malformed ROIs to be 1x1
+            ga_size roi_width = fmax((DTYPE_i0) (roi_end_w - roi_start_w + 1) * 1.0, 1);
+            ga_size roi_height =  fmax((DTYPE_i0) (roi_end_h - roi_start_h + 1) * 1.0, 1);
+
+            DTYPE_i0 bin_size_h = (DTYPE_i0)roi_height
+                             / (DTYPE_i0)pooled_height;
+            DTYPE_i0 bin_size_w = (DTYPE_i0)roi_width
+                             / (DTYPE_i0)pooled_width;
+
+            ga_size phstart = floor((DTYPE_i0)(h - roi_start_h) / bin_size_h);
```

Comment:
You need to subtract/add a small epsilon value within the floor/ceiling function to enure that boundary cases do not fail due to numerical precision.